group-coordinator/src/main/java/org/apache/kafka/coordinator/group/assignor/GeneralUniformAssignmentBuilder.java

```diff
@@ -14,17 +14,891 @@
+     * @param topicIdPartitions       The topic partitions that need to be sorted.
+     * @return A list of sorted topic partitions.
+     */
+    private List<TopicIdPartition> sortTopicIdPartitions(Collection<TopicIdPartition> topicIdPartitions) {
+        Comparator<TopicIdPartition> comparator = Comparator
+            .comparingDouble((TopicIdPartition topicIdPartition) -> {
+                int totalPartitions = subscribedTopicDescriber.numPartitions(topicIdPartition.topicId());
+                int totalSubscribers = membersPerTopic.get(topicIdPartition.topicId()).size();
+                return (double) totalPartitions / totalSubscribers;
+            })
+            .reversed()
+            .thenComparingInt(topicIdPartition -> membersPerTopic.get(topicIdPartition.topicId()).size())
+            .thenComparingInt(TopicIdPartition::partitionId);
+
+        return topicIdPartitions.stream()
+            .sorted(comparator)
+            .collect(Collectors.toList());
+    }
+
+    /**
+     * Gets a set of partitions that are to be retained from the existing assignment. This includes:
+     * <li> Partitions from topics that are still present in both the new subscriptions and the topic metadata. </li>
+     * <li> When using a rack-aware strategy, only partitions with member owners in the same rack are retained. </li>
+     * <li> Track current partition owners when there is a rack mismatch. </li>
+     */
+    private void assignStickyPartitions() {
+        members.forEach((memberId, assignmentMemberSpec) ->
+            assignmentMemberSpec.assignedPartitions().forEach((topicId, currentAssignment) -> {
+                if (assignmentMemberSpec.subscribedTopicIds().contains(topicId)) {
+                    currentAssignment.forEach(partition -> {
+                        TopicIdPartition topicIdPartition = new TopicIdPartition(topicId, partition);
+                        if (rackInfo.useRackStrategy && rackInfo.racksMismatch(memberId, topicIdPartition)) {
+                            currentPartitionOwners.put(topicIdPartition, memberId);
+                        } else {
+                            assignmentManager.maybeAssignPartitionToMember(topicIdPartition, memberId);
+                            assignedStickyPartitions.add(topicIdPartition);
+                        }
+                    });
+                } else {
+                    LOG.debug("The topic " + topicId + " is no longer present in the subscribed topics list");
+                }
+            })
+        );
+    }
+
+    /**
+     * Allocates the unassigned partitions to members in the same rack, if available.
+     */
+    private void rackAwarePartitionAssignment() {
+        // Sort partitions in ascending order by the number of potential members with matching racks.
+        // Only partitions with potential members in the same rack are returned.
+        List<TopicIdPartition> sortedPartitions = rackInfo.sortPartitionsByRackMembers(unassignedPartitions);
```

Comment:
this method sorts twice across all partitions. might need to look into this

Reply:
oh yes you're right, I've updated the method, thanks